### ISSUE TITLE
Incorrect output with a reused argument matcher

### ISSUE THREAD

**Issue by pirj** [rspec/rspec-expectations#1287]
### Subject of the issue

```ruby
RSpec.describe 'A', :aggregate_failures do
  specify do
    matcher = contain_exactly(eq(1))

    expect([2]).to matcher
    expect([3]).to matcher
  end
end
```

The second failure message seems to have traces from the first evaluation:
```
            the extra elements were:        [2]
```

It appears that `with_matchers_cloned` (`lib/rspec/matchers/composable.rb`) doesn't cover this case.

Adding `.clone` or `.dup` to `.to matcher` solves the issue.

### Your environment

* Ruby version: 2.7.1
* RSpec 3.10
  - rspec-core 3.10.1
  - rspec-expectations 3.10.1

### Expected behavior

```
  1) A is expected to contain exactly (eq 1)
     Got 2 failures:

     1.1) Failure/Error: expect([2]).to matcher.dup

            expected collection contained:  [(eq 1)]
            actual collection contained:    [2]
            the missing elements were:      [(eq 1)]
            the extra elements were:        [2]
          # ./spec/a_spec.rb:5:in `block (2 levels) in <top (required)>'

     1.2) Failure/Error: expect([3]).to matcher.dup

            expected collection contained:  [(eq 1)]
            actual collection contained:    [3]
            the missing elements were:      [(eq 1)]
            the extra elements were:        [3]
          # ./spec/a_spec.rb:6:in `block (2 levels) in <top (required)>'
```

### Actual behavior

```
  1) A is expected to contain exactly (eq 1)
     Got 2 failures:

     1.1) Failure/Error: expect([2]).to matcher

            expected collection contained:  [(eq 1)]
            actual collection contained:    [2]
            the missing elements were:      [(eq 1)]
            the extra elements were:        [2]
          # ./spec/a_spec.rb:5:in `block (2 levels) in <top (required)>'

     1.2) Failure/Error: expect([3]).to matcher

            expected collection contained:  [(eq 1)]
            actual collection contained:    [3]
            the missing elements were:      [(eq 1)]
            the extra elements were:        [2]
          # ./spec/a_spec.rb:6:in `block (2 levels) in <top (required)>'
```

**Comment by JonRowe** [comment]
Some output must be cached between runs, we are supposed to be allowing this use case.

**Comment by pirj** [comment]
Fixed in #1326
